### PR TITLE
feat: remove AWS SDK runtime dependency (SigV4 over fetch)

### LINKED ISSUE #36: Remove AWS SDK runtime dependency to enable n8n verification
## Summary

To qualify for n8n community-node **verification** (and installation on n8n Cloud), a package must ship with **no runtime dependencies** — the `@n8n/scan-community-package` static analysis must pass. This node currently depends on the AWS SDK:

- `@aws-sdk/client-bedrock-agentcore`
- `@aws-sdk/client-bedrock-agentcore-control`

The scanner lints every shipped `.js` file, and the AWS SDK's internal use of Node built-ins (`node:http`, `node:https`, `node:stream`, `process`, `setTimeout`) and its transitive `@smithy/*` packages fail the `no-restricted-imports` / `no-restricted-globals` rules. Bundling the SDK does not help — the violations are in the SDK code itself, and `eslint-disable` is ignored (`allowInlineConfig: false`).

## Proposed change

Remove the AWS SDK entirely and talk to the AgentCore control and data planes over HTTPS directly:

- **SigV4 request signing** implemented inline using only `node:crypto` (HMAC-SHA256 + SHA-256), which the scanner allows.
- **Control plane** (Create/Get/Update/Delete/List harness, versions, endpoints) via REST-JSON over `fetch`.
- **Data plane** `InvokeHarness` via `fetch` with the streaming `application/vnd.amazon.eventstream` response decoded by an inline event-stream parser (drops `@smithy/core/event-streams`).

Result: `dependencies` becomes empty, and `npx @n8n/scan-community-package` passes.

## Scope / guarantees

- **No user-facing change.** Same node, same `agentCoreApi` credential, same fields, same JSON output shape. Node `version` stays `1`. Existing workflows are unaffected.
- Also completes the licensing/packaging requirements for verification: relicense Apache-2.0 → MIT (approved), add author email, stop shipping type-declaration files.

## Related

Follows the verification-prep fixes in #28, #29, #30. This is the remaining (larger) blocker.

## Acceptance criteria

- [ ] `@aws-sdk/*` and `@smithy/*` removed from `dependencies`
- [ ] SigV4 signer validated against AWS test vectors and a live call
- [ ] Full flow retested against live AWS (create, memory multi-turn, tools, invoke streaming, VPC)
- [ ] `npx @n8n/scan-community-package` passes
- [ ] AppSec review of the inline SigV4 signer

### PR DESCRIPTION
**Pull request by sundargthb**
Closes #36.

## Summary

Removes the AWS SDK runtime dependency so the package can qualify for **n8n community-node verification** (verified nodes must ship with no runtime dependencies, which is also the gate for installing on n8n Cloud).

The two SDK clients are replaced with direct HTTPS calls:

- **Request signing** — a small inline **SigV4** signer using only `node:crypto` (HMAC-SHA256 + SHA-256), which the n8n scanner allows.
- **Control plane** (Create / Get / Update / Delete / List harness, versions, endpoints) — REST-JSON over `fetch`.
- **Data plane** `InvokeHarness` — `fetch` with the streaming `application/vnd.amazon.eventstream` response decoded by an inline event-stream parser, replacing `@smithy/core/event-streams`.

After this change `dependencies` is empty and `npx @n8n/scan-community-package` passes.

## No change to the UI or user experience

This is a **transport-layer change only**. There is intentionally no change to how the node looks or behaves:

- Same node, same `agentCoreApi` credential, same fields, same labels.
- Same output shape (`operation`, `harnessId`, `sessionId`, `harness` summary, `response`, `toolUses`, `usage`, `latencyMs`).
- Node `typeVersion` stays **1**, so existing workflows keep working unchanged.
- Same AWS APIs and the same IAM permissions as before (no new actions, no broadened scope).

## What changed

**New files**

| File | Purpose | Imports |
|---|---|---|
| `helpers/sigv4.ts` | AWS SigV4 request signer | `node:crypto` only |
| `helpers/eventstream.ts` | Binary event-stream frame decoder (inline CRC-32) | none |
| `helpers/httpClient.ts` | Sign + `fetch` + JSON parse + AWS error mapping | none |
| `helpers/controlClient.ts` | Control-plane operations as typed methods | none |

**Modified**

- `AgentCoreHarness.node.ts`, `helpers/client.ts`, `helpers/versioning.ts`, `helpers/oauth.ts` — call the new client/decoder instead of the SDK. Business logic (provisioning, drift detection, tools/memory/skills/model builders, session handling) is unchanged.
- `package.json` — `dependencies` emptied; SDK moved to `devDependencies` (used only by the offline validation script); `license` set to `MIT`; author email added.
- `LICENSE` — Apache-2.0 to MIT (approved separately).
- `tsconfig.json` — stop emitting `.d.ts` / `.map` (not needed at runtime; keeps the published artifact clean for the scanner).
- SPDX headers updated to `MIT` across source.

**Validation script**

- `scripts/poc-getharness.mjs` — offline parity check that signs a `GetHarness` call two ways (new signer vs. the SDK) and compares. Not shipped in the published package.

## Verification and testing

- `npm run build`, `npm run typecheck`, `npm run lint`, `npm run format:check` all pass.
- `npx @n8n/scan-community-package` passes (previously failed on SDK-internal use of Node built-ins).
- SigV4 signer validated against the published AWS SigV4 test vector and against a live `GetHarness` call (byte-for-byte match with the SDK response).
- Full flow retested against live AWS:
  - Run Agent (auto-provision) with provided Session ID, multi-turn memory recall.
  - Code Interpreter tool run (nested tool-use JSON reconstructed correctly from the event stream).
  - Invoke Existing Harness (BYO ARN).
  - Credential Test (control-plane `ListHarnesses`).

## Security review

The two new files (`sigv4.ts`, `eventstream.ts`) introduce request signing and untrusted-input binary parsing, and are undergoing a security review. IAM permissions, credential handling, TLS behavior, and the `no-eval` invariant are unchanged; runtime dependency count is reduced from two to zero.

## Follow-ups (not in this PR)

- Version bump / release will be handled via the release workflow once review completes.
- README install/verification wording will be revisited once verification lands.

### CODE REVIEW

**Comment by sundargthb** [comment]
Added a unit test suite in this PR (per review feedback that the repo lacked tests).

**Framework:** vitest, added as a dev-dependency only — production dependency count stays zero. Tests live under `test/` and are excluded from the published `dist/`. A `test` job now runs in the Quality and Safety Checks workflow.

**Coverage (69 tests, 8 files):**
- `sigv4.test.ts` — canonical-algorithm determinism, plus a differential test asserting byte-identical Authorization headers versus `@smithy/signature-v4` (the AWS SDK's own signer) across path/query/unicode/ARN-in-query edge cases; confirms the secret key never appears in signed output.
- `eventstream.test.ts` — happy-path frame decode, multi-frame and split-chunk reassembly, exception-frame shaping, and safe rejection of corrupt prelude/message CRCs and oversized frames (exercises the T-N8N-017 mitigations directly).
- `tools`, `memory`, `model`, `skills` builders + `configHash` drift detection.
- `stream` accumulator (text, streamed tool-use JSON reassembly, error events).
- `httpClient` request signing + AWS error mapping (fetch mocked).

This turns the ad-hoc validation done for the signer and decoder into committed regression tests that run on every change. Broader integration tests against live AWS remain a follow-up.

**Review by akhtrma-amz — APPROVED** [review]

File: nodes/AgentCoreHarness/helpers/httpClient.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+	const res = await fetch(url, {
+		method: req.method,
+		headers,
+		...(bodyString ? { body: bodyString } : {}),
+	});
+
+	const text = await res.text();
+	if (!res.ok) {
+		throw new Error(formatAwsError(res.status, res.statusText, text));
```
Comment (notgitika) [inline nodes/AgentCoreHarness/helpers/httpClient.ts]: AWS REST-JSON errors often carry the type only in `x-amzn-errortype`. Dropping the headers turns a missing endpoint into a generic 404, so the upsert path will not create it. Could we check the header, body `code`, then `__type`?

File: nodes/AgentCoreHarness/helpers/eventstream.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+				}
+				if (buffer.length < totalLength) break; // frame not fully arrived yet
+
+				const frame = buffer.subarray(0, totalLength);
+				const event = decodeFrame(frame);
+				if (event) yield event;
+
+				buffer = buffer.subarray(totalLength);
+			}
+		}
```
Comment (notgitika) [inline nodes/AgentCoreHarness/helpers/eventstream.ts:90]: Could we throw if EOF leaves bytes buffered? A truncated response currently exits normally, so `consumeStream` can return a partial result as a successful invocation.

File: nodes/AgentCoreHarness/helpers/httpClient.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		{
+			method: req.method,
+			url,
+			headers: { 'content-type': 'application/json' },
+			body: bodyString,
+		},
+		{ region: config.region, service: SERVICE, credentials: config.credentials },
+	);
+
+	const res = await fetch(url, {
```
Comment (notgitika) [inline nodes/AgentCoreHarness/helpers/httpClient.ts]: This drops the SDK’s standard retries, so one transient network, throttling, or 5xx failure now aborts the operation. Could we add bounded retries and preserve one `clientToken` per mutating request?

File: nodes/AgentCoreHarness/helpers/httpClient.ts (hunk 1 above)
Comment (sundargthb) [reply nodes/AgentCoreHarness/helpers/httpClient.ts]: Good catch, this was a real bug. formatAwsError now reads the type from the x-amzn-errortype header first, then a body code field, then __type. The endpoint upsert keys off ResourceNotFoundException, so with the header-only case it was silently becoming a plain 404 and skipping the create. Fixed and covered by a test that returns the type only in the header.

File: nodes/AgentCoreHarness/helpers/eventstream.ts (hunk 2 above)
Comment (sundargthb) [reply nodes/AgentCoreHarness/helpers/eventstream.ts:90]: Agreed. The decoder now throws at EOF if any bytes remain buffered, since a well-formed stream ends on a frame boundary and leftover bytes mean the response was cut off mid-frame. That turns a truncated response into an error instead of consumeStream returning a partial result as a success. Added a test that feeds a full frame plus a partial one and expects it to throw.

File: nodes/AgentCoreHarness/helpers/httpClient.ts (hunk 3 above)
Comment (sundargthb) [reply nodes/AgentCoreHarness/helpers/httpClient.ts]: Done. controlRequest now retries transient failures (429, 500, 502, 503, 504, and network errors) with exponential backoff, capped at 4 attempts. Each mutating request carries a clientToken that stays the same across retries, so a replayed write is idempotent server side. All the harness create/update/delete and endpoint ops accept clientToken, so this applies everywhere it matters. Tests cover retry-then-succeed, no-retry on 400, giving up after the cap, and the token staying stable across attempts.

**Review by notgitika — APPROVED** [review]

**Review by tynobleamazon — APPROVED** [review]
